### PR TITLE
Throw ConverterException properly for missing numerical value

### LINKED ISSUE #869: Resource Indexer Exception in causes observations sync download to fail.
**Describe the bug**
Resource Indexer has issue extracting Temperature QuantityValue in the following Observation taken from the Synthea dataset.

```
{
	"resourceType": "Observation",
	"id": "1595",
	"meta": {
		"versionId": "1",
		"lastUpdated": "2021-10-21T08:29:32.481+00:00",
		"source": "#RqWIwZVJSul7Y3tp"
	},
	"status": "final",
	"category": [{
		"coding": [{
			"system": "http://terminology.hl7.org/CodeSystem/observation-category",
			"code": "vital-signs",
			"display": "vital-signs"
		}]
	}],
	"code": {
		"coding": [{
			"system": "http://loinc.org",
			"code": "8331-1",
			"display": "Oral temperature"
		}],
		"text": "Oral temperature"
	},
	"subject": {
		"reference": "Patient/1561"
	},
	"encounter": {
		"reference": "Encounter/1593"
	},
	"effectiveDateTime": "2011-09-15T12:22:44-04:00",
	"issued": "2011-09-15T12:22:44.702-04:00",
	"valueQuantity": {
		"value": 37.65532501487887,
		"unit": "Cel",
		"system": "http://unitsofmeasure.org",
		"code": "Cel"
	}
}
```

**Exception:**

```
10-26 12:21:42.804 20937 20978 W System.err: java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String org.fhir.ucum.Decimal.asDecimal()' on a null object reference
10-26 12:21:42.805 20937 20978 W System.err: 	at com.google.android.fhir.UnitConverter.getCanonicalForm$engine_debug(UnitConverter.kt:47)
10-26 12:21:42.805 20937 20978 W System.err: 	at com.google.android.fhir.index.ResourceIndexer.quantityIndex(ResourceIndexer.kt:307)
10-26 12:21:42.805 20937 20978 W System.err: 	at com.google.android.fhir.index.ResourceIndexer.extractIndexValues(ResourceIndexer.kt:96)
10-26 12:21:42.805 20937 20978 W System.err: 	at com.google.android.fhir.index.ResourceIndexer.index(ResourceIndexer.kt:68)
10-26 12:21:42.805 20937 20978 W System.err: 	at com.google.android.fhir.db.impl.dao.ResourceDao.insertResource(ResourceDao.kt:202)
10-26 12:21:42.805 20937 20978 W System.err: 	at com.google.android.fhir.db.impl.dao.ResourceDao.insertAll$suspendImpl(ResourceDao.kt:74)
10-26 12:21:42.805 20937 20978 W System.err: 	at com.google.android.fhir.db.impl.dao.ResourceDao.insertAll(Unknown Source:4)
10-26 12:21:42.805 20937 20978 W System.err: 	at com.google.android.fhir.db.impl.DatabaseImpl.insertRemote(DatabaseImpl.kt:68)
10-26 12:21:42.805 20937 20978 W System.err: 	at com.google.android.fhir.db.impl.DatabaseImpl$insertSyncedResources$2.invokeSuspend(DatabaseImpl.kt:97)
10-26 12:21:42.805 20937 20978 W System.err: 	at com.google.android.fhir.db.impl.DatabaseImpl$insertSyncedResources$2.invoke(Unknown Source:12)
10-26 12:21:42.805 20937 20978 W System.err: 	at com.google.android.fhir.db.impl.DatabaseImpl$insertSyncedResources$2.invoke(Unknown Source:7)
10-26 12:21:42.805 20937 20978 W System.err: 	at androidx.room.RoomDatabaseKt$withTransaction$2.invokeSuspend(RoomDatabase.kt:58)
10-26 12:21:42.805 20937 20978 W System.err: 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
10-26 12:21:42.805 20937 20978 W System.err: 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
10-26 12:21:42.805 20937 20978 W System.err: 	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:274)
10-26 12:21:42.805 20937 20978 W System.err: 	at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:85)
10-26 12:21:42.805 20937 20978 W System.err: 	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59)
10-26 12:21:42.806 20937 20978 W System.err: 	at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source:1)
10-26 12:21:42.806 20937 20978 W System.err: 	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38)
10-26 12:21:42.806 20937 20978 W System.err: 	at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source:1)
10-26 12:21:42.806 20937 20978 W System.err: 	at androidx.room.RoomDatabaseKt$acquireTransactionThread$$inlined$suspendCancellableCoroutine$lambda$2.run(RoomDatabase.kt:121)
10-26 12:21:42.806 20937 20978 W System.err: 	at androidx.room.TransactionExecutor$1.run(TransactionExecutor.java:47)
10-26 12:21:42.806 20937 20978 W System.err: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
10-26 12:21:42.806 20937 20978 W System.err: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
10-26 12:21:42.806 20937 20978 W System.err: 	at java.lang.Thread.run(Thread.java:923)
10-26 12:21:42.809 20937 20972 D FhirSynchronizer: download ended at 1635231102809, Tue Oct 26 12:21:42 GMT+05:30 2021, total time 11397 ms

```

### LINKED ISSUE #974: Follow up for #952
apologies this got merged after @aditya-07 approved this because I turned on auto-merge... will put this into my next pr (and tag you)

_Originally posted by @jingtang10 in https://github.com/google/android-fhir/pull/952#discussion_r762624924_

### PR DESCRIPTION
**Pull request by jingtang10**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #869 
Fixes #974 

**Description**

- Throw ConverterException properly for missing numerical value
- Also closes #974 which handles the follow-up comments in #952 

**Alternative(s) considered**
NA

**Type**
Choose one: Bug fix

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md)
- [x] I have read [How to Contribute](https://github.com/google/android-fhir/blob/master/docs/contributing.md)
- [x] I have read the [Developer's guide](https://github.com/google/android-fhir/wiki/Developer's-Guide)
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate )
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally
- [x] I have built and run the reference app(s) to verify my change fixes the issue and/or does not break the reference app(s)

### CODE REVIEW

File: common/src/main/java/com/google/android/fhir/UnitConverter.kt [hunk 1]
```diff
@@ -47,13 +56,14 @@ object UnitConverter {
         pair.code,
         pair.value.asDecimal().toBigDecimal(MathContext(value.value.precision()))
       )
-    } catch (exception: UcumException) {
-      exception.printStackTrace()
-      throw ConverterException(exception)
+    } catch (e: UcumException) {
+      throw ConverterException("UCUM conversion failed", e)
+    } catch (e: NullPointerException) {
+      throw ConverterException("Missing numerical value in the canonical UCUM value", e)
```
Comment (epicadk) [inline common/src/main/java/com/google/android/fhir/UnitConverter.kt:62]: Should we add the link to the issue here?

File: engine/src/main/java/com/google/android/fhir/search/filter/TokenParamFilterCriterion.kt [hunk 2]
```diff
@@ -33,7 +33,7 @@ import org.hl7.fhir.r4.model.UriType
  * of(CodeType("male")) })
  */
 @SearchDslMarker
-data class TokenParamFilterCriterion internal constructor(var parameter: TokenClientParam) :
+data class TokenParamFilterCriterion constructor(var parameter: TokenClientParam) :
```
Comment (aditya-07) [inline engine/src/main/java/com/google/android/fhir/search/filter/TokenParamFilterCriterion.kt:36]:
```suggestion
data class TokenParamFilterCriterion(var parameter: TokenClientParam) :
```

**Review by aditya-07 — CHANGES_REQUESTED** [review]

File: common/src/main/java/com/google/android/fhir/UnitConverter.kt (hunk 1 above)
Comment (jingtang10) [reply common/src/main/java/com/google/android/fhir/UnitConverter.kt:63]: done

File: engine/src/main/java/com/google/android/fhir/search/filter/TokenParamFilterCriterion.kt (hunk 2 above)
Comment (jingtang10) [reply engine/src/main/java/com/google/android/fhir/search/filter/TokenParamFilterCriterion.kt]: done

**Review by jingtang10 — COMMENTED** [review]
thanks both!

**Review by aditya-07 — APPROVED** [review]
LGTM

**Pull request merged**